c/storage_mem.c

```diff
@@ -0,0 +1,286 @@
+  const char *valueCopy = memStorageCopyString(storage, value);
+  if (!keyCopy || !valueCopy) {
+    if (keyCopy) {
+      memStorageFree((char*)keyCopy);
+    }
+    if (valueCopy) {
+      memStorageFree((char*)valueCopy);
+    }
+    *statusOut = STORAGE_ALLOC_ERROR;
+    return;
+  }
+  int status = memStorageLock(storage);
+  if (status) {
+    *statusOut = status;
+    return;
+  }
+  htPut(storage->table, (void*)keyCopy, (void*)valueCopy);
+  status = memStorageUnlock(storage);
+  *statusOut = status;
+}
+
+static const char *memStorageGetString(MemStorage *storage, const char *key, int *statusOut) {
+  int status = memStorageLock(storage);
+  if (status) {
+    *statusOut = status;
+    return NULL;
+  }
+  const char *value = htGet(storage->table, (void*)key);
+  bool found = (value != NULL);
+  status = memStorageUnlock(storage);
+  *statusOut = !found ? STORAGE_KEY_NOT_FOUND : status;
+  return value;
+}
+
+static void memStorageRemove(MemStorage *storage, const char *key, int *statusOut) {
+  int status = memStorageLock(storage);
+  if (status) {
+    *statusOut = status;
+    return;
+  }
+  int removed = htRemove(storage->table, (void*)key);
+  status = memStorageUnlock(storage);
+  *statusOut = !removed ? STORAGE_KEY_NOT_FOUND : status;
+}
+
+static const char *MESSAGES[] = {
+  [STOARGE_LOCK_ERROR] = "Failed to lock storage",
+  [STOARGE_UNLOCK_ERROR] = "Failed to unlock storage",
+  [STORAGE_ALLOC_ERROR] = "Failed to allocate memory",
+};
+
+#define MESSAGE_COUNT sizeof(MESSAGES)/sizeof(MESSAGES[0])
+
+static const char *memStorageGetStrStatus(MemStorage *storage, int status) {
+  int adjustedStatus = status - STORAGE_FIRST_CUSTOM_STATUS;
+  if (adjustedStatus >= MESSAGE_COUNT || adjustedStatus < 0) {
+    return "Unknown status code";
+  }
+  const char *message = MESSAGES[adjustedStatus];
+  if (!message) {
+    return "Unknown status code";
+  }
+  return message;
+}
+
+Storage *makeMemoryStorage(MemoryStorageOptions *options) {
+  Storage *storage = (Storage*)safeMalloc(sizeof(*storage), "Storage");
+  if (!storage) {
+    return NULL;
+  }
+  MemStorage *memStorage = makeMemStorage(options);
+  if (!memStorage) {
+    safeFree((char*)storage, sizeof(*storage));
+    return NULL;
+  }
+  storage->userData = memStorage;
+  storage->set = (StorageSet) memStorageSetString; 
+  storage->get = (StorageGet) memStorageGetString; 
+  storage->remove = (StorageRemove) memStorageRemove;
+  storage->strStatus = (StorageGetStrStatus) memStorageGetStrStatus;
+  return storage;
+}
+
+#ifdef _TEST
+
+/*
+
+Build:
+
+c89 \
+  -D_TEST=1 \
+  -D_XOPEN_SOURCE=600 \
+  -D_OPEN_THREADS=1 \
+  -Wc,langlvl\(extc99\),gonum,goff,hgpr,roconst,ASM,asmlib\('CEE.SCEEMAC','SYS1.MACLIB','SYS1.MODGEN'\) \
+  -Wc,agg,exp,list\(\),so\(\),off,xref \
+  -I ../h \
+  -o ./storage \
+  alloc.c \
+  collections.c \
+  storage.c \
+  storage_mem.c \
+  utils.c \
+  timeutls.c 
+
+Run: ./storage
+
+*/
+
+int main(int argc, char *argv[]) {
```

Comment:
Could you do something like
* Put a value to the store
* Check if the value is in the store using `assert`
* Check a non-existent value in the store using `assert` (should not be there)
* etc

That way you'll see right away if something unexpected happens.